### DISCUSSION TITLE
X3-MIC/PRO-G2

### DISCUSSION THREAD

**Discussion by horamarques** [wills106/homeassistant-solax-modbus#130] [General]
hi

do you have any news about the support of X3-MIC/PRO-G2?
I've just connected my inverter but the available entities is very small

do you plan to add more features for this inverter?

Thanks

**Comment by wills106** [comment]
They are all the usefull registers that I know off.

**Comment by horamarques** [comment]
After your feedback I realised that my integration is not working.
I Have an X3-MIC/PRO-G2 that as a meter connected. From the manual available online I used pins 1&2 to connect my waveshare RS485 to Ethernet. 
but it seems that on this generation of X3 there are only pins 4&5 available to connect using RS485.
![image](https://user-images.githubusercontent.com/11884874/194855376-c18edc6d-7309-4a19-b797-5ed33c985a0c.png)

Is this correct?

thanks again for your help.

regards

**Reply by wills106** [reply]
You normally have separate ports for the Meter and the RS485. Pins 4&5 are the ones you want for your Waveshare in that picture.

**Reply by horamarques** [reply]
but on those I have the meter connected...
it seems that on this version they changed and there's no option to have a meter and RS485 working...

**Reply by wills106** [reply]
If you check on the LCD on the Inverter you should be able to see what address your Inverter is set to. It might not be the standard address of 1

**Reply by horamarques** [reply]
Hi, yes I'm using address1 on the inverter.

Regarding RS485 wiring, I got feedback from solax:
"there are no additional RS485 ports on the inverter, if you want to connect a second meter, please also connect to the same RS485 ports on the inverter, just make sure the two meters are in different addresses"

My first attempt with this config did not work, I was not able to get data on HomeAssistant and the inverter lost the connection to the meter.

any recommendation on this setup?

thanks

**Comment by horamarques** [comment]
I Have all the wiring done, and seem to be correct, at leats the meter is still working...
but I have this error on HA and the device never gets ready:
Logger: homeassistant.config_entries
Source: custom_components/solax_modbus/plugin_solax.py:507 
Integration: SolaX Inverter Modbus ([documentation](https://github.com/wills106/homsassistant-solax-modbus), [issues](https://github.com/wills106/homsassistant-solax-modbus/issues)) 
First occurred: 09:18:02 (1 occurrences) 
Last logged: 09:18:02

Error setting up entry SolaX for solax_modbus
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 365, in async_setup
    result = await component.async_setup_entry(hass, self)
  File "/config/custom_components/solax_modbus/__init__.py", line 104, in async_setup_entry
    plugin_path = config.get(CONF_PLUGIN, DEFAULT_PLUGIN)
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/config/custom_components/solax_modbus/plugin_solax.py", line 507, in <module>
    SolaxModbusNumberEntityDescription( name = "Export Control User Limit",
TypeError: SolaxModbusNumberEntityDescription.__init__() got an unexpected keyword argument 'read_scale_exceptions'

any idea?

thanks

**Reply by infradom** [reply]
read_scale_exceptions is something that has been introduced in beta code a day ago.
Did you install from Github or a beta ?
Apologies for the inconvenience, should be solved soon ..

**Reply by infradom** [reply]
The code on github should be consistent (today). Dont know why your version's const.py is not in sync with plugin_solax.py

**Reply by horamarques** [reply]
I'm using the last beta version.
I removed and added the device again, and I'm back to the initial point... not able to identify the inverter module...

any help you can provide to investigate connection issues between waveshare and the inverter? Im trying to use the Multi-host... since I only have 1 rs485 connection on the inverter

**Comment by infradom** [comment]
I have no experience with MIC, but it would surprise me that the RS485 modbus port would be the same as the meter's modbus port.
On my hybrid X3 G4, this is a completely different port 
Multi host is meant for another scenario (multipe HA stations), so I would not activate that

**Reply by infradom** [reply]
Furthermore, connecting to pin 1 and 2 may have destroyed your waveshare ..

**Comment by infradom** [comment]
From a MIC manual:
<img width="554" alt="image" src="https://user-images.githubusercontent.com/11804014/195833940-62d3dd36-8e4f-4bc8-8275-fdb0ac110a7b.png">

It could indeed be the same RJ45 plug, but the meter connects to pin7-8 and waveshare to 4-5

If this is correct, pin 1 and 2 are not used and your waveshare will not be destroyed .. ;-)

Remember also polarity matters: dont swap the A and B lead (it wont destroy your equipment however)

**Reply by horamarques** [reply]
The feedbackI got from solax is that my inverter PINs are:
![image](https://user-images.githubusercontent.com/11884874/195836625-01ff8eb0-01a4-48e5-805a-3862e9251a1f.png)

so the only option is to connect on the same pins...

**Reply by infradom** [reply]
I tend to not believe that if you have a MIC, but who am I to pretend I know better ...
Anyway, if they are right and you connected it once to pin1-2, your waveshare may be damaged

**Reply by infradom** [reply]
What is your model number?  My picture is form the manual for models 
X3-8.0P-T-D X3-8.0P-T-N
X3-10.0P-T-D X3-10.0P-T-N
X3-12.0P-T-D X3-12.0P-T-N
X3-15.0P-T-D X3-15.0P-T-N

**Reply by horamarques** [reply]
I have a X3-MIC-4k-G2
And the feedback from solax was:
"Your message is well received, for your question below, there are no additional RS485 ports on the inverter, if you want to connect a second meter, please also connect to the same RS485 ports on the inverter, just make sure the two meters are in different addresses."

**Reply by wills106** [reply]
I think they are referring to a second Modbus power meter like a Eastron's SDM630?

**Reply by infradom** [reply]
I found the manual for X3-MIC-4k-G2 and indeed, the same RS485 bus is used for meter and monitoring/control.
My apologies. 
In this case, I hope the 12V on pins1-2 have not damaged your waveshare (unlikely but possible)

**Comment by horamarques** [comment]
Back to testing this integration

I think that the wiring is now well done and communication is working.
using pymodbus im able to get values from the waveshare, but still not able to make it work on HA. From the logs it's not able to detect the inverter type. 
Any ideia on how to move forward? any additional test I cloud do with pymodbus?

Thanks in advance

**Reply by horamarques** [reply]
temperature seems to be working (some times) 
![image](https://user-images.githubusercontent.com/11884874/205053005-ea0c849e-2857-43da-82d2-393a2cb82981.jpeg)

**Reply by wills106** [reply]
I have just realised you own a 4kW X3 Mic, Is that correct? If so the Serial number isn't In the Integration so wont work. What's the first 4 digits? I can add it in and then it should work, once I release a new Beta

**Reply by horamarques** [reply]
My serial is: MC204TI…

thanks for you help

**Reply by wills106** [reply]
If you give 0.7.0b62 ago it should work.

**Reply by PeterPin71** [reply]
> Ak dáte pred 0.7.0b62 malo by to fungovať.

Hi Wills106. Please help me. I have Solax X3 MIC/PRO inventer. I would like to connect it to home assisatnt via modbus. But your  plugin_solax.py doesn't support my converter. Could we somehow get it into the plugin so that it communicates? very very thank for your help. S/N my inverter starts MU702THB....

**Reply by wills106** [reply]
Just released [2023.02.2](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2023.02.2) with support for Serial's starting MC702T

**Reply by PeterPin71** [reply]
great. thank you very much.

**Reply by PeterPin71** [reply]
MC and MU are the same?

**Reply by wills106** [reply]
Sorry I misread the serial number you provided.
[2023.02.3](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2023.02.3) includes MU702T

**Comment by horamarques** [comment]
Still no luck

![B053B557-D94A-405E-84E5-9C4DA3C6652D](https://user-images.githubusercontent.com/11884874/205159847-f8dcafc0-7259-4861-b11c-9561301d7cd0.png)

**Reply by wills106** [reply]
If it's saying unknown you don't have working communications to the Inverter.

**Reply by horamarques** [reply]
Any idea on How to debug it?
On what address should I get the serial and how many bytes?

**Reply by wills106** [reply]
If you read holding register 0x300 for a count of 7 as a string, you should get your serial number.

**Reply by horamarques** [reply]
Im not able to read the register at 0x300, but it works fine on 0x0

![image](https://user-images.githubusercontent.com/11884874/205708743-22bd2d19-ca32-456b-a31f-8a84b296fc2f.jpeg)


on 0x300 no matter what length I always have 

![image](https://user-images.githubusercontent.com/11884874/205708681-c2f598a9-a03f-46b3-98d4-cc53f45a293b.jpeg)


I already checked the inverter address and its 1. Communication seems to be working, correct?
On 0x8 I have:

![image](https://user-images.githubusercontent.com/11884874/205708648-9be1107d-1aa7-445b-8425-90ef3a0e74bb.jpeg)

**Reply by horamarques** [reply]
Any additional ideia to debug this?

**Comment by michaelbeelman** [comment]
Hello all, I'm new to this forum, but am on and off working the last three years to read my Solax X3-MIC inverter locally, without the use of Solax Cloud, to no avail so far.. This post has been super helpful and clarifying I many ways, I feel I just miss a simple thing here.

![solaxx3micrs485usb](https://user-images.githubusercontent.com/122157338/211150148-502069c1-1b74-40de-8604-00d7cc1f35c6.png)

I run Home Assistant 2022.12.8 in a Docker container using docker-compose.yml file, on a server running Ubuntu 22.10. This seems to run fine with lots of integrations (e.g. Z-Wave). I am using the proposed RS485-USB adapter using the blue-white blue wires that connect to the RS485 port on the Solax X3 MIC (ports 4 and 5 for 485_A and 485_B respectively).

I can confirm I have the ch341 UART running on /dev/ttyUSB0. 

<img width="519" alt="ttuUSB0" src="https://user-images.githubusercontent.com/122157338/211150266-0581c0e9-1923-4dec-af8e-1f8072d9165a.png">

This is what I get in Node-RED: "Error, port not open!".

<img width="849" alt="Scherm­afbeelding 2023-01-07 om 13 19 46" src="https://user-images.githubusercontent.com/122157338/211150307-c4664f84-882d-4daa-b79a-89f8fd48982d.png">

And my settings:

<img width="531" alt="Scherm­afbeelding 2023-01-07 om 13 20 05" src="https://user-images.githubusercontent.com/122157338/211150311-b06a4bfa-2001-4e58-b47b-207520155d29.png">

Now, when I try with Modbus tool from the command line, I get timed-out response. 

<img width="781" alt="Scherm­afbeelding 2023-01-07 om 13 21 20" src="https://user-images.githubusercontent.com/122157338/211150384-2d63c910-05b7-49cb-af94-e3d0d9508f65.png">

Now, disabling all of the above, installing your Solax Inverter Modbus custom-integration and after a cold-restart of the server incl. HA, I still don't get anything. What could I be missing here?

<img width="522" alt="Scherm­afbeelding 2023-01-07 om 13 46 09" src="https://user-images.githubusercontent.com/122157338/211151337-63a00d48-d4d1-4a83-90d4-6826460da611.png">

<img width="377" alt="Scherm­afbeelding 2023-01-07 om 13 46 16" src="https://user-images.githubusercontent.com/122157338/211151343-3822ed81-eae4-4dd2-9717-8cd5c174a054.png">

No SolaX Inverter Temperature readings.. none :(

Here are the logs:

<img width="1235" alt="Scherm­afbeelding 2023-01-07 om 13 49 19" src="https://user-images.githubusercontent.com/122157338/211151472-dd83d877-8991-40a5-8acf-ce0cfc91e1ca.png">

Here is info on my inverter: 
- Machine Type: X3-5.0
- Master: 1.10
- Slave: 1.02
- Manager: 1.09

![WhatsApp Image 2023-01-07 at 13 57 21-2](https://user-images.githubusercontent.com/122157338/211151782-8671c9f6-f03f-40ac-a48f-6506e7558219.jpeg)
![WhatsApp Image 2023-01-07 at 13 57 21](https://user-images.githubusercontent.com/122157338/211151783-c9746456-0d1a-414f-bf8a-717bc8b43bdf.jpeg)
![WhatsApp Image 2023-01-07 at 13 57 20](https://user-images.githubusercontent.com/122157338/211151784-7ff6b3b3-10d2-4a05-a13f-faf9ebd7994a.jpeg)

I really hope someone will be able to guide on next steps on how to make this work. Hopefully this way I can contribute to make this integration a little more complete by proving it is working on Solax X3 inverter!

**Comment by infradom** [comment]
We really need the detailed HA logs to help you. There must be messages since our integration is quite verbose during startup. (make sure info, warning and error messages are enabled) (so recommended logging level = INFO)

I personnaly have always avoided node-red so I cannot help you with that.

**Reply by infradom** [reply]
Correction, sorry I see the logs, overlooked it

**Reply by michaelbeelman** [reply]
This is what I get in the docker logs. What I find strange is that even outside of HA, like in Node-RED or even bare-bone via the command-line, I'm not able to get it to work, even though it just comes down to two wires into a USB stick. I can put a nice RJ45 connector on both ends and put a tester on it just to see if the wiring is proper.

<img width="1384" alt="Scherm­afbeelding 2023-01-07 om 14 10 45" src="https://user-images.githubusercontent.com/122157338/211152343-1969559d-b474-4df4-bbcd-914b41ce7f94.png">

**Reply by infradom** [reply]
These errors point to wiring,  baudrate errors or device address errors
I am surprised to see "reading input registers at device 2" 
It should be using device address 1 normally if your screenshot of the initial configuration is correct.
Most inverters have a default modbus address=1

**Reply by michaelbeelman** [reply]
Yes, I tried 1-10, to no avail. I have put it back to 1. I will check the wiring once more and do a full power cycle of the inverter unit afterwards.

**Reply by infradom** [reply]
In your latest screenshot, I see even device address 10 very weird.

We dont like using usb rs485 adapters, due to the "blocking call" warnings, but that is not the cause of your problems

**Reply by wills106** [reply]
I would double check your wiring.
With RS485 A goes to A and B goes to B.
They don't need swapping over like you would with a Serial Null Modem cable.

**Reply by infradom** [reply]
Indeed, no swapping allowed

**Reply by michaelbeelman** [reply]
I will confirm shortly! Your feedback is much appreciated.

**Reply by michaelbeelman** [reply]
I have found a buggy cable connection. I made a new wire and now the serial number is recognized via RS485-USB adapter into Home Assistant: MC502TF4172046. Though it still fails reading input registers at device 1 position 0x8. It also fails to read Inverter Temperature, even after 5 minutes.

<img width="1238" alt="Scherm­afbeelding 2023-01-07 om 14 44 43" src="https://user-images.githubusercontent.com/122157338/211153893-c96cee5f-7ef2-4d3c-bc69-ae2f62a46995.png">

**Reply by wills106** [reply]
Your Inverter isn't supported (yet), that's why you have "unrecognized inverter type - serial number :" error.
I'll add you serial in and push an update out.

Edit: It's an older Gen1 Inverter isn't it?
Edit 2: Noticed you mentioned trying for 3 years. So I am confident it's a Gen1

**Reply by wills106** [reply]
Give [2023.01.2b6](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2023.01.2b6) a try.

**Reply by michaelbeelman** [reply]
Works like a charm! Many, many thanks for your expert support and also quick response times. The result is there. Another Gen1 X3 device supported by your integration. What's the best place to make a donation to support your work?

<img width="343" alt="Scherm­afbeelding 2023-01-07 om 15 22 04" src="https://user-images.githubusercontent.com/122157338/211155416-1515489f-41aa-4683-8a33-38678cdf986b.png">

**Comment by jtrefil** [comment]
Hello,
I am trying to enable communication via wifi module ver3 with X3-MIC-15.0K-T-D. I previously solved the communication with the X3-HYBRID G4 and without problems.
I'll try to flash the FW wifi dongle v.3, but I'd rather throw this question in as well.
The installation is a bit far from me so I need to check a few things before I go physically.
sn: MC215THC305002; MC215THC305011
Thank you very much for the kick.
Jan

**Reply by wills106** [reply]
I don't have the serials MC215T in the Integration yet, but I can add them.
Do you know if these are Gen1 or Gen2 X3-Mic's?

**Reply by jtrefil** [reply]
this is a G2, thanks for adding.

**Reply by wills106** [reply]
Ok, I should get them added in later today and a new release sent out.

**Reply by jtrefil** [reply]
thanks a lot...

**Reply by wills106** [reply]
Just released [2023.02.5](https://github.com/wills106/homeassistant-solax-modbus/releases/tag/2023.02.5) should be working now.

**Reply by Trilis29** [reply]
Does communication via wifi module working?

**Comment by Trilis29** [comment]
Has anyone managed to connect smart meter and this plugin on same wires?

**Reply by wills106** [reply]
Yes, that's how you connect the X3-Mic G2. The Modbus power Meter and connection to PC/Server is on the same RS485 connection.

**Reply by raenji-sk** [reply]
@wills106 do you have any reference or more details about simultaneous connection to HA and meter? How does it work then, who is the client and who is the server? I mean can the inverter act as both being server for HA and client for meter?

One more thing, I recently updated the WiFi dongle FW. It now sends modbus data too,. I was able to conect it to NodeRed so I assume it should work with your integration too yes? Leaving the COM/Meter port on Solax Mic G2 free to use for meter or "emulator".

**Comment by Roving-Ronin** [comment]
Hi,

Would anyone happen to know if a Solax X3-PRO G2 (S/N Format: MPT10TIC###### ) is compatible with this integration, and if so what Dongle works best (reading notes it appears to be the LSE-3 WiFi/Ethernet) ?

Apologies not a Solax user myself, so not familiar with the models/in's and outs, trying to assist a mate who's been sold one of these (no dongle or consumption meter either) and trying to help him as he wants to get it 'usable' and able to access data (i.e. HA + SunSynk card).

Many Thanks in Advance !

**Comment by nta629** [comment]
Hello all,

I am trying to connect my inverter using this integration but no matter how hard I try I can't. I have an X3 12.0P-T-D inverter, this inverter has one rs485 port, on pins 4 and 5 I have Waveshare connected to communicate with HA and on pins 7 and 8 I have a DTSU666-CT meter. I have tried so many settings options but I am left clueless. The only setting that found / added about 20 entities for me was using the solax_lv version but it couldn't read the values ​​from them. I am just starting out with HA so I apologize for the lack of necessary information but I will be happy to provide any information you may require from me.

**Reply by raenji-sk** [reply]
Got an X3 MIC G2 with one RS485 port. Should work fine via WiFi and WiFi-Lan dongle v3 with latest firmware if you have those. Been using it like that for over a year without any issues. RS485 is wired to meter, sort of, and WiFi is used for integration data.
